### PR TITLE
Dog and fix features

### PR DESCRIPTION
**Pull request by ducha-aiki**
### Description
- Changes ScalePyramid for local feature detection so that it matches OpenCV scale pyramid, especially for DoG (SIFT) detector.
- Added Difference of Gaussian detector
- Fixes a bug in ConvQuad3D scale estimation output.

!! Breaking change: now ScalePyramid returns shape of [B, Ch, Layers, H, W] as opposed to previous [B, Levels, Ch, H, W].
I believe, that scale dimension is more natural to be treated as scale dimension, so conv3d, pool3d could be directly applicable

### Status
Ready

### Types of changes

- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x ] **Pass all tests**: did you test in local ? `make test`
- [x ] Unittests: did you add tests for your new functionality ?
- [x ] Documentations: did you build documentation ? `make build-docs`
- [x ] Implementation: is your code well commented and follow conventions ? `make lint`
- [x ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [x] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [x] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by ducha-aiki** [comment]
@edgarriba 

Locally I have 

> =================================================================================================== mypy ===================================================================================================
Success: no issues found in 68 source files

Any suggestions what should I update locally?

**Comment by edgarriba** [comment]
@ducha-aiki maybe yes. I found in the past some issue with Tuple type. I guess the reason is padding, and I think I fixed it by using List instead.

**Comment by ducha-aiki** [comment]
Python 3.8 fails tests with weird error: 

  File "/home/circleci/kornia-tester/tutorials/filter_edges.py", line 10, in <module>
    import cv2
ImportError: /home/circleci/kornia-tester/.dev_env/envs/venv/lib/python3.8/site-packages/../../././libgnutls.so.30: symbol mpn_add_1 version HOGWEED_4 not defined in file libhogweed.so.4 with link time reference

File: kornia/feature/laf.py [hunk 1]
```diff
@@ -477,21 +477,26 @@ def extract_patches_from_pyramid(img: torch.Tensor,
...
-def laf_is_inside_image(laf: torch.Tensor, images: torch.Tensor) -> torch.Tensor:
+def laf_is_inside_image(laf: torch.Tensor, images: torch.Tensor,
+                        b: int = 0) -> torch.Tensor:
     """Checks if the LAF is touching or partly outside the image boundary. Returns the mask
     of LAFs, which are fully inside the image, i.e. valid.
 
     Args:
         laf (torch.Tensor):  :math:`(B, N, 2, 3)`
         images (torch.Tensor): images, lafs are detected in :math:`(B, CH, H, W)`
+        border (int): additional border
```
Comment (edgarriba) [inline kornia/feature/laf.py:488]: argument name should match with function signature `b != border`

File: test/geometry/transform/test_crop.py [hunk 2]
```diff
@@ -40,6 +40,10 @@ def test_bounding_boxes_dim_inferring_batch(self, device):
         h, w = kornia.geometry.transform.crop._infer_bounding_box(boxes)
         assert (h, w) == (expected_height, expected_width)
 
+    @pytest.mark.skip(reason="Crashes with pytorch internal error")
+    #  RuntimeError: isDifferentiableType(variable.scalar_type()) INTERNAL ASSERT FAILED at
+    # "/opt/conda/conda-bld/pytorch_1595629417679/work/torch/csrc/autograd/functions/utils.h":59,
+    # please report a bug to PyTorch.
```
Comment (edgarriba) [inline test/geometry/transform/test_crop.py:46]: is this for pt 1.6.0 already ?

Comment (ducha-aiki) [reply test/geometry/transform/test_crop.py:46]: yes

**Pull request merged**